src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala

```diff
@@ -22,281 +22,77 @@
-
-  private val refreshDelay : Long = (10 * 60 * 1000) + rnd.nextInt(5 * 60 * 1000) // in 10 - 15 minutes
-  private val refreshPeriod : Long = 15 * 60 * 1000 // every 15 minutes
-
-  // main purpose of the time is to allow bulk operations to consume
-  // additional throughput when more RUs are getting provisioned
-  private val timerName = "collectionthroughput-refresh-timer"
-  private val timer: Timer = new Timer(timerName)
-
-  @transient private var bulkImporter: DocumentBulkExecutor = _
-
-  private var documentClient: DocumentClient = CosmosDBConnection.getClient(connectionMode, getClientConfiguration(config))
-
-  logInfo(s"${timerName}: scheduling timer - delay: ${refreshDelay} ms, period: ${refreshPeriod} ms")
-  timer.schedule(
-    new TimerTask { def run() = {
-      collectionThroughput = None
-      logInfo(s"${timerName}: Resetting cached collection throughput")
-    }},
-    refreshDelay, 
-    refreshPeriod
-  )
-
-  def getPartitionKeyDefinition(partitionKeyDefinition: Option[String]) : PartitionKeyDefinition = {
-    if (partitionKeyDefinition.isDefined) {
-        val pkDefinition = new PartitionKeyDefinition()
-        val paths: ListBuffer[String] = new ListBuffer[String]()
-        paths.add(partitionKeyDefinition.get)
-        pkDefinition.setPaths(paths)  
-
-        pkDefinition
-    }
-    else {
-      getCollection.getPartitionKey
-    }
-  }
+  private val clientConfig = ClientConfiguration(config)
 
-  def getDocumentBulkImporter(collectionThroughput: Int, partitionKeyDefinition: Option[String], maxMiniBatchUpdateCount: Int, maxMiniBatchImportSizeKB: Int): DocumentBulkExecutor = {
-    if (bulkImporter == null) {
-      val initializationRetryOptions = new RetryOptions()
-      initializationRetryOptions.setMaxRetryAttemptsOnThrottledRequests(1000)
-      initializationRetryOptions.setMaxRetryWaitTimeInSeconds(1000)
-
-      val pkDefinition = getPartitionKeyDefinition(partitionKeyDefinition)
-
-      bulkImporter = DocumentBulkExecutor.builder.from(documentClient,
-        databaseName,
-        collectionName,
-        pkDefinition,
-        collectionThroughput
-      ).withInitializationRetryOptions(initializationRetryOptions)
-        .withMaxUpdateMiniBatchCount(maxMiniBatchUpdateCount)
-        .withMaxMiniBatchSize(maxMiniBatchImportSizeKB * 1024).build()
-    }
-
-    bulkImporter
-  }
-
-  def setDefaultClientRetryPolicy: Unit = {
-    if (documentClient != null) {
-      documentClient.getConnectionPolicy().getRetryOptions().setMaxRetryAttemptsOnThrottledRequests(1000);
-      documentClient.getConnectionPolicy().getRetryOptions().setMaxRetryWaitTimeInSeconds(1000);
-    }
-  }
-
-  def setZeroClientRetryPolicy: Unit = {
-    if (documentClient != null) {
-      documentClient.getConnectionPolicy().getRetryOptions().setMaxRetryAttemptsOnThrottledRequests(0);
-      documentClient.getConnectionPolicy().getRetryOptions().setMaxRetryWaitTimeInSeconds(0);
-    }
+  def getCollectionLink: String = {
```

Comment:
Does this and the lazy val still need to be here?  Quick scan of the class I see some methods take collectionLink in as a parameter, others don't.  I suspect the ones that do would call this method first anyway.

Reply:
I could retrieve it within the methods form the cache. But the lazy Val is safe here - the self link is immutable - so keeping it here won't cause issues.